pkg/provider/cloud/openstack/provider.go

```diff
@@ -314,7 +325,8 @@ func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, updat
 	// reconciliations so far. This is to avoid hitting OpenStack API at each
 	// iteration.
 	// TODO: this is terrible, find a better way.
-	if cluster.Status.ExtendedHealth.CloudProviderInfrastructure != kubermaticv1.HealthStatusUp {
+	if cluster.Status.ExtendedHealth.CloudProviderInfrastructure != kubermaticv1.HealthStatusUp &&
+		kubernetes.HasFinalizer(cluster, RouterSubnetLinkCleanupFinalizer) {
```

Comment:
Isn't this the wrong way around? Attaching the subnet if the finalizer is already set?

Reply:
it's not because finalizers are added ahead 
https://github.com/kubermatic/kubermatic/blob/master/pkg/provider/cloud/openstack/provider.go#L187-L199